staking/slash/double-sign.go

```diff
@@ -1,52 +1,341 @@
+	return string(s)
+}
+
+// MarshalJSON ..
+func (r Record) MarshalJSON() ([]byte, error) {
+	reporter, offender :=
+		common2.MustAddressToBech32(r.Reporter),
+		common2.MustAddressToBech32(r.Offender)
+	return json.Marshal(struct {
+		Evidence         Evidence `json:"evidence"`
+		Beneficiary      string   `json:"beneficiary"`
+		AddressForBLSKey string   `json:"offender"`
+	}{r.Evidence, reporter, offender})
+}
+
+func (e Evidence) String() string {
+	s, _ := json.Marshal(e)
+	return string(s)
+}
+
+func (r Record) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
+
+// CommitteeReader ..
+type CommitteeReader interface {
+	ReadShardState(epoch *big.Int) (*shard.State, error)
+}
 
 // Verify checks that the signature is valid
-func Verify(candidate *Record) error {
+func Verify(chain CommitteeReader, candidate *Record) error {
+	first, second :=
+		candidate.Evidence.AlreadyCastBallot,
+		candidate.Evidence.DoubleSignedBallot
+	if shard.CompareBlsPublicKey(first.SignerPubKey, second.SignerPubKey) != 0 {
+		k1, k2 := first.SignerPubKey.Hex(), second.SignerPubKey.Hex()
+		return errors.Wrapf(
+			errBLSKeysNotEqual, "%s %s", k1, k2,
+		)
+	}
+	superCommittee, err := chain.ReadShardState(candidate.Evidence.Epoch)
+
+	if err != nil {
+		return err
+	}
+
+	subCommittee := superCommittee.FindCommitteeByID(
+		candidate.Evidence.ShardID,
+	)
+
+	if subCommittee == nil {
+		return errors.Wrapf(
+			errShardIDNotKnown, "given shardID %d", candidate.Evidence.ShardID,
+		)
+	}
+
+	if _, err := subCommittee.AddressForBLSKey(second.SignerPubKey); err != nil {
+		return err
+	}
+	// TODO need to finish this implementation
 	return nil
 }
 
-// Apply ..
-func Apply(state *state.DB, slashes []byte) error {
+var (
+	errBLSKeysNotEqual = errors.New(
+		"bls keys in ballots accompanying slash evidence not equal ",
+	)
+	errSlashDebtNotFullyAccountedFor = errors.New(
+		"slash debt was not fully accounted for, still non-zero",
+	)
+	errShardIDNotKnown              = errors.New("nil subcommittee for shardID")
+	errValidatorNotFoundDuringSlash = errors.New("validator not found")
+	zero                            = numeric.ZeroDec()
+	oneDoubleSignerRate             = numeric.MustNewDecFromStr("0.02")
+	oneHundredPercent               = numeric.NewDec(1)
+	fiftyPercent                    = numeric.MustNewDecFromStr("0.5")
+)
+
+// applySlashRate returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
+func applySlashRate(amount *big.Int, rate numeric.Dec) *big.Int {
+	return numeric.NewDecFromBigInt(
+		amount,
+	).Mul(rate).TruncateInt()
+}
+
+func delegatorSlashApply(
+	snapshot, current *staking.ValidatorWrapper,
+	rate numeric.Dec,
+	state *state.DB,
+	reporter common.Address,
+	doubleSignEpoch *big.Int,
+	slashTrack *Application,
+) error {
+	for _, delegationSnapshot := range snapshot.Delegations {
+		slashDebt := applySlashRate(delegationSnapshot.Amount, rate)
+		halfOfSlashDebt := new(big.Int).Div(slashDebt, common.Big2)
```

Comment:
Again, this halfOfSlashDebt should the half of what's actually taken away from the stake, not the calculated amount to pay, because they may not have enough to pay.